src/main/java/io/quarkus/gizmo/Type.java

```diff
@@ -0,0 +1,418 @@
+package io.quarkus.gizmo;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.PrimitiveType.Primitive;
+
+/**
+ * This interface can be used to build a JVM signature for classes, methods and fields.
+ * <p>
+ * Implementations are instantiated via factory methods; for example {@link #voidType()} and {@link #classType(Class)}.
+ * 
+ * @see SignatureBuilder
+ */
+public interface Type {
+
+    // factory methods
+
+    public static VoidType voidType() {
+        return VoidType.INSTANCE;
+    }
+
+    public static PrimitiveType byteType() {
+        return PrimitiveType.BYTE;
+    }
+
+    public static PrimitiveType booleanType() {
+        return PrimitiveType.BOOLEAN;
+    }
+
+    public static PrimitiveType intType() {
+        return PrimitiveType.INT;
+    }
+
+    public static PrimitiveType longType() {
+        return PrimitiveType.LONG;
+    }
+
+    public static PrimitiveType shortType() {
+        return PrimitiveType.SHORT;
+    }
+
+    public static PrimitiveType doubleType() {
+        return PrimitiveType.DOUBLE;
+    }
+
+    public static PrimitiveType floatType() {
+        return PrimitiveType.FLOAT;
+    }
+
+    public static PrimitiveType charType() {
+        return PrimitiveType.CHAR;
+    }
+
+    public static ClassType classType(DotName name) {
+        return new ClassType(Objects.requireNonNull(name).toString().replace('.', '/'));
+    }
+
+    public static ClassType classType(String name) {
+        return new ClassType(name);
+    }
+
+    public static ClassType classType(Class<?> classType) {
+        Objects.requireNonNull(classType);
+        if (classType.getDeclaringClass() != null) {
+            return classType(classType.getDeclaringClass().getName().replace('.', '/') + "." + classType.getSimpleName());
+        }
+        return classType(classType.getName().replace('.', '/'));
+    }
+
+    public static ParameterizedType parameterizedType(ClassType classType, Type... typeArguments) {
+        return new ParameterizedType(classType, Arrays.asList(typeArguments));
+    }
+
+    public static ArrayType arrayType(Type component) {
```

Comment:
This is not a _component_ type, but an _element_ type. There's no difference between these terms in single-dimensional arrays, but in e.g. `String[][]`, the component type is `String[]`, while the element type is `String`.

Reply:
Ok, I got confused by the `jakarta.enterprise.lang.model.types.ArrayType.componentType()` javadoc ;-). By the way, `org.jboss.jandex.ArrayType.component` is then wrong too.

Reply:
The CDI Lang Model javadoc has the exact same example :-)

The Jandex `ArrayType.component()` is generally an element type, but may sometimes be a component type, which is confusing and probably needs even more documentation than I already added. Jandex has more quirks like this, unfortunately.